### PR TITLE
RavenDB-21321 AVE in encrypted index

### PR DESCRIPTION
**Pull request by ayende**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-21321

### Additional description

We had a couple of places where we missed the `ModifyPage` call. It means that we modified the page *outside of the transaction context*.
That is *really* bad, what is worse is that it will _usually_ work on non encrypted database.
On encrypted database, that will cause the data to by modified on the _current_ transaction and go away on the _next_ transaction.

Added *lots* of debug asserts to catch this, and added additional validation via the encrypted pager to catch this in more general terms.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change
### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

### Testing by RavenDB QA team

- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.

*We need to re-test the encrypted tests*

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Voron/Data/Lookups/Lookup.cs [hunk 1]
```diff
@@ -423,6 +424,7 @@ private bool MaybeMergeEntries(ref CursorState destinationState)
     {
         Debug.Assert(_internalCursor._pos > 0, "_internalCursor._pos > 0");
         ref var parent = ref _internalCursor._stk[_internalCursor._pos - 1];
+        parent.Page = _llt.ModifyPage(parent.Page.PageNumber);
```
Comment (ayende) [inline src/Voron/Data/Lookups/Lookup.cs:427]: This is the actual fix for the AVE

File: src/Voron/Data/PostingLists/PostingList.cs [hunk 2]
```diff
@@ -432,7 +432,7 @@ public void PrepareForCommit()
                     var siblingIdx = GetSiblingIndex(parent);
                     var (_, siblingPageNum) = branch.GetByIndex(siblingIdx);
 
-                    var siblingPage = _llt.GetPage(siblingPageNum);
+                    var siblingPage = _llt.ModifyPage(siblingPageNum);
```
Comment (ayende) [inline src/Voron/Data/PostingLists/PostingList.cs:435]: This is another case that _might_ cause this

**Comment by ayende** [comment]
test this in debug please

**Comment by redknightlois** [comment]
LGTM

**Comment by ayende** [comment]
test in debug please

**Review by maciejaszyk — APPROVED** [review]

**Comment by arekpalinski** [comment]
test this in debug please

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**